### PR TITLE
ARROW-8655: [C++][Python] Preserve partitioning information for a discovered Dataset

### PR DESCRIPTION
**Pull request by jorisvandenbossche**

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
See the JIRA for more details and motivating use cases, but a small demo of what this exposes:

```python
import pandas as pd
import pyarrow as pa
import pyarrow.parquet as pq
import pyarrow.dataset as ds

df = pd.DataFrame({"year": [2020, 2020, 2021, 2021], "month":[1, 2, 1, 2], "values": [1, 2, 3, 4]})
df.to_parquet("test_partitioned", partition_cols=["year", "month"], engine="pyarrow")
```

A discovered dataset now stores the Partitioning object that was created/discovered in the FileSystemDatasetFactory, so you can still access this information on the dataset:

```python
>>> dataset = ds.dataset("test_partitioned/", partitioning="hive")
>>> dataset
<pyarrow._dataset.FileSystemDataset at 0x7fe481b6e270>
>>> dataset.partitioning
<pyarrow._dataset.HivePartitioning at 0x7fe48162c7f0>

# the schema (field names and types) of the partitioning
>>> dataset.partitioning.schema 
year: int32
month: int32
>>> dataset.partitioning.schema.names
['year', 'month']
# and all partition field values discovered during the factory
>>> dataset.partitioning.dictionaries
[<pyarrow.lib.Int32Array object at 0x7fe480fd9fa0>
 [
   2020,
   2021
 ],
 <pyarrow.lib.Int32Array object at 0x7fe480fd9b80>
 [
   1,
   2
 ]]
```

With inferring a dictionary type for the partitioning, the schema of the partitioning object indeed has dictionary type:

```python
>>> dataset2 = ds.dataset("test_partitioned/", partitioning=ds.HivePartitioning.discover(infer_dictionary=True))
>>> datatset2.partitioning.schema
year: dictionary<values=int32, indices=int32, ordered=0>
month: dictionary<values=int32, indices=int32, ordered=0>
-- schema metadata --
pandas: '{"index_columns": [], "column_indexes": [{"name": null, "field_n' + 352
```

When the FileSystemDataset is not discovered with a partitioning or not created through a discovery, the `partitioning` attribute is None:

```python
>>> dataset3 = ds.dataset("test_partitioned/")
# TODO this is still something to fix, right now this is a "default" partitioning (which isn't exposed in Python)
>>> dataset3.partitioning
...
~/scipy/repos/arrow/python/pyarrow/_dataset.pyx in pyarrow._dataset.Partitioning.wrap()
TypeError: default

>>> dataset4 = ds.FileSystemDataset(list(dataset.get_fragments()), dataset.schema, dataset.format, dataset.filesystem)
>>> dataset4.partitioning is None
True
```

File: cpp/src/arrow/dataset/file_base.h [hunk 1]
```diff
@@ -238,6 +238,12 @@ class ARROW_DS_EXPORT FileSystemDataset : public Dataset {
       std::shared_ptr<FileFormat> format, std::shared_ptr<fs::FileSystem> filesystem,
       std::vector<std::shared_ptr<FileFragment>> fragments);
 
+  static Result<std::shared_ptr<FileSystemDataset>> Make(
+      std::shared_ptr<Schema> schema, compute::Expression root_partition,
+      std::shared_ptr<FileFormat> format, std::shared_ptr<fs::FileSystem> filesystem,
+      std::vector<std::shared_ptr<FileFragment>> fragments,
+      std::shared_ptr<Partitioning> partitioning);
```
Comment (bkietz) [inline cpp/src/arrow/dataset/file_base.h:245]: In the interest of limiting available constructors, please combine these overloads:
```suggestion
      std::vector<std::shared_ptr<FileFragment>> fragments,
      std::shared_ptr<Partitioning> partitioning = NULLPTR);
```

File: cpp/src/arrow/dataset/partition.h [hunk 2]
```diff
@@ -172,6 +172,8 @@ class ARROW_DS_EXPORT KeyValuePartitioning : public Partitioning {
 
   Result<std::string> Format(const compute::Expression& expr) const override;
 
+  const ArrayVector dictionaries() const { return dictionaries_; }
```
Comment (bkietz) [inline cpp/src/arrow/dataset/partition.h:175]:
```suggestion
  const ArrayVector& dictionaries() const { return dictionaries_; }
```

File: cpp/src/arrow/dataset/discovery.cc [hunk 3]
```diff
@@ -274,7 +274,8 @@ Result<std::shared_ptr<Dataset>> FileSystemDatasetFactory::Finish(FinishOptions
     fragments.push_back(fragment);
   }
 
-  return FileSystemDataset::Make(schema, root_partition_, format_, fs_, fragments);
+  return FileSystemDataset::Make(schema, root_partition_, format_, fs_, fragments,
+                                 partitioning);
```
Comment (bkietz) [inline cpp/src/arrow/dataset/discovery.cc:278]:
```suggestion
  return FileSystemDataset::Make(std::move(schema), root_partition_, format_, fs_,
                                 std::move(fragments), std::move(partitioning));
```

**Review by bkietz — CHANGES_REQUESTED** [review]
This seems fine, just a few nits:

File: python/pyarrow/_dataset.pyx [hunk 4]
```diff
@@ -2083,6 +2090,15 @@ cdef class DirectoryPartitioning(Partitioning):
         return PartitioningFactory.wrap(
             CDirectoryPartitioning.MakeFactory(c_field_names, c_options))
 
+    @property
+    def dictionaries(self):
```
Comment (westonpace) [inline python/pyarrow/_dataset.pyx:2094]: Perhaps add a docstring / comment here explaining that this property might be empty or null if the partitioning was not created through some kind of discovery process?  For example:

```
# partitioning.dictionaries will be empty here
partitioning = HivePartitioning(pa.schema([("year", pa.int16()), ("month", pa.int8())]))
```

File: python/pyarrow/_dataset.pyx [hunk 5]
```diff
@@ -2214,6 +2230,15 @@ cdef class HivePartitioning(Partitioning):
         return PartitioningFactory.wrap(
             CHivePartitioning.MakeFactory(c_options))
 
+    @property
+    def dictionaries(self):
```
Comment (westonpace) [inline python/pyarrow/_dataset.pyx:2234]: Same comment here.

**Review by westonpace — APPROVED** [review]
This seems like a good approach.  If the partitioning was created by discovery the dictionaries should be present.

**Comment by westonpace** [comment]
I'll add my comment from the JIRA here as you suggested.  Another approach could be to expose the dictionaries as part of the partitioning factory output.  So the partitioning factory could have a `dictionaries()` property that is set after `Finish` has been called.  This puts a bit more complexity on python since it hides the partitioning factory from the user but it could track the dictionaries and expose them where needed.

File: cpp/src/arrow/dataset/file_base.h (hunk 1 above)
Comment (jorisvandenbossche) [reply cpp/src/arrow/dataset/file_base.h:245]: Yes, I still wanted to ask about that if I could use a default null argument ;) Will edit this

File: python/pyarrow/_dataset.pyx (hunk 4 above)
Comment (jorisvandenbossche) [reply python/pyarrow/_dataset.pyx:2106]: Yes, added some more docstrings now

File: python/pyarrow/tests/test_dataset.py [hunk 6]
```diff
@@ -1999,39 +2003,46 @@ def test_scan_iterator(use_threads, use_async):
...
+
+    full_table = table.append_column(
+        "part", pa.array(np.repeat([0, 1, 2], 3), type=pa.int32()))
+
+    return full_table, path
+
+
+@pytest.mark.parquet
+def test_open_dataset_partitioned_directory(tempdir, dataset_reader):
+    full_table, path = _create_partitioned_dataset(tempdir)
```
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_dataset.py:2026]: Those changes are unrelated, but edited the test a bit to factor out the test data creation in a `_create_partitioned_dataset` helper function to reuse.

**Comment by jorisvandenbossche** [comment]
Updated the PR according to the comments + added tests and docstrings.

>  Another approach could be to expose the dictionaries as part of the partitioning factory output. So the partitioning factory could have a `dictionaries()` property that is set after `Finish` has been called. This puts a bit more complexity on python since it hides the partitioning factory from the user but it could track the dictionaries and expose them where needed.

In the end, I am fine either way *how* we achieve it, and I suppose this could also work (in the python factory functions, we pass the PartitioningFactory before calling `Finish` on the dataset factory, so we have access to this object, could extract dictionaries, and attach this to the Dataset class only at the cython level). 
If this would be preferred, I can take a look at it. But, I think the changes on the C++ side in the current implementation are rather limited and not invasive, and this makes this easier at the Python side. I also don't know to what extent there would be interest in this on the R side (it could make a read CSV -> write parquet flow easier for the case where you want to preserve the exact partitioning).

File: cpp/src/arrow/dataset/file_base.h [hunk 7]
```diff
@@ -236,7 +236,8 @@ class ARROW_DS_EXPORT FileSystemDataset : public Dataset {
   static Result<std::shared_ptr<FileSystemDataset>> Make(
       std::shared_ptr<Schema> schema, compute::Expression root_partition,
       std::shared_ptr<FileFormat> format, std::shared_ptr<fs::FileSystem> filesystem,
-      std::vector<std::shared_ptr<FileFragment>> fragments);
+      std::vector<std::shared_ptr<FileFragment>> fragments,
+      std::shared_ptr<Partitioning> partitioning = NULLPTR);
```
Comment (bkietz) [inline cpp/src/arrow/dataset/file_base.h:240]: This parameter needs a doccomment https://github.com/apache/arrow/pull/10661/checks?check_run_id=3019338484#step:8:2650

**Review by bkietz — CHANGES_REQUESTED** [review]

File: python/pyarrow/_dataset.pyx [hunk 8]
```diff
@@ -2083,6 +2102,27 @@ cdef class DirectoryPartitioning(Partitioning):
...
+        """
+        cdef vector[shared_ptr[CArray]] c_arrays
+        c_arrays = self.directory_partitioning.dictionaries()
+        res = []
+        for arr in c_arrays:
+            if arr.get() == nullptr:
+                # Partitioning object has not been created through
+                # inspected Factory
+                return None
+            res.append(pyarrow_wrap_array(arr))
```
Comment (bkietz) [inline python/pyarrow/_dataset.pyx:2123]: Probably not worth worrying about here (will not occur in any KeyValuePartitioning constructed during discovery) but it *is* possible for a KeyValuePartitioning to have dictionaries for some fields and not others

**Review by bkietz — APPROVED** [review]
LGTM

File: python/pyarrow/_dataset.pyx (hunk 8 above)
Comment (jorisvandenbossche) [reply python/pyarrow/_dataset.pyx:2123]: `pyarrow_wrap_array` will in that case raise an error (it checks for `arr` being NULL), so at least it won't segfault I think

**Pull request closed without merging**